### ISSUE TITLE
Two Double value comparision - "Floating point numbers should not be tested for equality"

### ISSUE THREAD

**Issue by jagatgame** [SonarSource/sonar-dotnet#3325]
### Description

Hi, I am facing issue in Sonar synced with SonarQube developer edition 8.1 

- Floating point numbers should not be tested for equality (csharpsquid:S1244)

### Repro steps

`//example code
if (Math.Round(val1) == Math.Round(val2)) {
}`
val1 & val2 is double type & Math.Round returns double type too
### Expected behavior

Should be allowed to compare above way? or else I am not aware of another way to compare the double values 

### Related information

* C# 7.0
* Visual Studio Version - 2017 Enterprise
* Win 10
* Sonar Qube Developer edition 8.1
* Sonar Lint add on latest

**Comment by pavel-mikula-sonarsource** [comment]
Hello @jagatgame,

Internal representation of `double` floating point type is implemented as mantissa and exponent under the hood all the way down to CPU/FPU. That brings precision risks of size defined by double.Epsilon` constant.

The true native precise way of doing this would look like this. This also works for rounding to different decimal position.
```C#
var rounded1 = Math.Round(val1);
var rounded2 = Math.Round(val2);
if (rounded1 - double.Epsilon <= rounded2 && rounded2 <= rounded1 + double.Epsilon)
{
}
```
To express what you really need (to compare integer representation of the number), you can
```C#
if ((int)Math.Round(val1) == (int)Math.Round(val2))
{
}
```
Or, knowing that Convert.ToInt32 rounds the number, you can
```C#
if (Convert.ToInt32(val1) == Convert.ToInt32(val2))
{
}
```
Second and third solution works for rounding to different decimal position as well, when multiplied accordingly.
